### PR TITLE
Ready - Improved hydration validation

### PR DESCRIPTION
**Pull request by sbarker2**
Merged:  https://github.com/atlassian-labs/nadel/pull/461/files/e3ff70e3eb4a60f03f1fed2a6e014dc2775de069

This PR is Ready For Review and not awaiting any other PRs 👍 

Tested in central schema ✅ 
Passed Engine Tests ✅ 
Added unit tests covering all cases ✅

### CODE REVIEW

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+import graphql.schema.*
+
+internal class NadelHydrationArgumentValidation private constructor() {
+    companion object {
+        /*
+*  Checks the type of a hydration argument derived from either a $source field output type or $argument field input
+* argument type against the type of an actor field argument to see if they match
+* will also allow for cases of batched hydration eg ID! -> [ID]
+*/
+        fun getHydrationInputErrors(
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:19]: Have you considered batch hydration here? It's possible to have `users(ids: [ID!]!): [User!]` be the actor field but have `$source.userId` that references a `userId: ID` i.e. `[ID!]!` vs `ID` but it is still compatible.

For basic hydration this looks good.

Nit: please run the IntelliJ formatter on the code :)

Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:19]: can you elaborate on that example above so I can make a test case?
Currently if we have the required input as `[ID!]!` but supply just an `ID` then it will fail validation. Why are they supposed to be compatible in the case of batch hydration?

I don't see how batch vs non-batch makes a different to the test since we are just checking the input types

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            val unwrappedHydrationSourceFieldType = hydrationSourceFieldType.unwrapNonNull() as GraphQLObjectType
+            val unwrappedActorFieldArgType = actorFieldArgType.unwrapNonNull() as GraphQLInputObjectType
+
+            val hydrationInnerFields: Map<String, GraphQLType> = unwrappedHydrationSourceFieldType.fields.associate { it.name to it.type as GraphQLType }
+            val actorInnerFields: Map<String, GraphQLType> = unwrappedActorFieldArgType.fields.associate { it.name to it.type as GraphQLType }
+
+            for (innerField in actorInnerFields){
+                val actorInnerFieldName = innerField.key
+                val actorInnerFieldType = innerField.value
+                val hydrationType = hydrationInnerFields.get(actorInnerFieldName)
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:149]:
```suggestion
            for (actorInnerField in unwrappedActorFieldArgType.fields) {
                val actorInnerFieldName = actorInnerField.name
                val actorInnerFieldType = actorInnerField.type
                val hydrationType = unwrappedHydrationSourceFieldType.getField(actorField.name)?.type
```

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                        hydrationSourceFieldType,
+                        actorFieldArgType,
+                )
+            }
+            return null
+        }
+
+        private fun validateInputObjectArg(
+                hydrationSourceFieldType: GraphQLType,
+                actorFieldArgType: GraphQLType,
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:134]:
```suggestion
                hydrationSourceFieldType: GraphQLObjectType,
                actorFieldArgType: GraphQLInputObjectType,
```

Per above comment about passing in smart casted objects

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+                parent: NadelServiceSchemaElement,
+                overallField: GraphQLFieldDefinition,
+                remoteArg: RemoteArgumentDefinition,
+                hydration: UnderlyingServiceHydration
+        ): NadelSchemaValidationError? {
+            val unwrappedHydrationSourceFieldType = hydrationSourceFieldType.unwrapNonNull() as GraphQLObjectType
+            val unwrappedActorFieldArgType = actorFieldArgType.unwrapNonNull() as GraphQLInputObjectType
+
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:142]:
```suggestion
```

Per above comment about passing in smart casted objects

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+            // check the actual type
+            if (unwrappedHydrationSourceFieldType is GraphQLScalarType && unwrappedActorFieldArgType is GraphQLScalarType) {
+                return validateScalarArg(hydrationSourceFieldType, actorFieldArgType, parent, overallField, remoteArg, hydration)
+            }
+            else if (unwrappedHydrationSourceFieldType.isList && unwrappedActorFieldArgType.isList) {
+                return validateListArg(hydrationSourceFieldType, actorFieldArgType, parent, overallField, remoteArg, hydration)
+            }
+            else if (unwrappedHydrationSourceFieldType is GraphQLObjectType && unwrappedActorFieldArgType is GraphQLInputObjectType) {
+                return validateInputObjectArg(hydrationSourceFieldType, actorFieldArgType, parent, overallField, remoteArg, hydration)
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:51]:
```suggestion
                return validateInputObjectArg(unwrappedHydrationSourceFieldType, unwrappedActorFieldArgType, parent, overallField, remoteArg, hydration)
```

You should pass in the smart casted objects instead of passing in the original values then manually unwrapping and casting them.

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            val unwrappedActorFieldArgType = actorFieldArgType.unwrapNonNull() as GraphQLInputObjectType
+
+            val hydrationInnerFields: Map<String, GraphQLType> = unwrappedHydrationSourceFieldType.fields.associate { it.name to it.type as GraphQLType }
+            val actorInnerFields: Map<String, GraphQLType> = unwrappedActorFieldArgType.fields.associate { it.name to it.type as GraphQLType }
+
+            for (innerField in actorInnerFields){
+                val actorInnerFieldName = innerField.key
+                val actorInnerFieldType = innerField.value
+                val hydrationType = hydrationInnerFields.get(actorInnerFieldName)
+                if (hydrationType == null) {
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:150]: Mentioned offline that we need to check if the input field is required or not.

File: test/src/test/kotlin/graphql/nadel/tests/EngineTests.kt [hunk 7]
```diff
@@ -35,7 +35,7 @@ import java.util.concurrent.CompletableFuture
  * 2. Test name e.g. hydration inside a renamed field
  * 3. Copy paste output from selecting a test in the IntelliJ e.g. java:test://graphql.nadel.tests.EngineTests.current hydration inside a renamed field
  */
-private val singleTestToRun = (System.getenv("TEST_NAME") ?: "")
+private val singleTestToRun = (System.getenv("TEST_NAME") ?: "basic hydration")
```
Comment (bbakerman) [inline test/src/test/kotlin/graphql/nadel/tests/EngineTests.kt:38]: I think this is bad - it will run only that test instead of all of them ??

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt (hunk 6 above)
Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:214]: done, thanks

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt (hunk 1 above)
Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt]: added the two missing cases as discussed offline:

```
//could have ID feed into [ID] (as a possible batch hydration case).
            // in this case we need to unwrap the list and check types
            if (isBatchHydration && (!unwrappedHydrationSourceFieldType.isList && unwrappedActorFieldArgType.isList)) {
```
and
```
//could have [ID] feed into ID (non-batch, ManyToOne case explained in NadelHydrationStrategy.kt)
            // in this case we need to unwrap the list and check types
            else if (!isBatchHydration && (unwrappedHydrationSourceFieldType.isList && !unwrappedActorFieldArgType.isList)) {
```

File: test/src/test/kotlin/graphql/nadel/tests/EngineTests.kt (hunk 7 above)
Comment (sbarker2) [reply test/src/test/kotlin/graphql/nadel/tests/EngineTests.kt:38]: whoops, was not supposed to push that

File: test/src/test/kotlin/graphql/nadel/tests/CentralSchemaTesting.kt [hunk 8]
```diff
@@ -48,7 +48,7 @@ const val runQuery = false
  */
 suspend fun main() {
     val schema = File(
-        "/Users/fwang/Documents/Atlassian/graphql-central-schema/schema/",
+        "/Users/sbarker2/git/graphql-central-schema/schema/",
```
Comment (sbarker2) [inline test/src/test/kotlin/graphql/nadel/tests/CentralSchemaTesting.kt:51]: todo: undo this change

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationValidation.kt [hunk 9]
```diff
@@ -227,14 +239,38 @@ internal class NadelHydrationValidation(
...
+                            hydration,
+                            isBatchHydration
+                    )
                 }
             }
 
-            else -> {
-                null
+            StaticArgument -> {
+                val staticArg = remoteArgSource.staticValue
```
Comment (sbarker2) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationValidation.kt:257]: wondering if this logic should be moved to `NadelHydrationArgumentValidation` like the other cases

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationValidation.kt [hunk 10]
```diff
@@ -227,14 +239,38 @@ internal class NadelHydrationValidation(
...
+                if (!validationUtil.isValidLiteralValue(
```
Comment (sbarker2) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationValidation.kt:258]: isValidLiteralValue is a public recursive function from this `graphql-java` class:
https://github.com/graphql-java/graphql-java/blob/dd50648b5e178f965918d0c352787bd81654c7fe/src/main/java/graphql/validation/ValidationUtil.java

This is an internal class, but I have checked over the code and confirmed it is what we want, including recursively checking inner types of objects/functions and everything.
Have also added extensive tests for static args to check this in NadelHydrationArgumentValidationTest.kt
It is an internal class in

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 11]
```diff
@@ -88,8 +90,9 @@ internal class NadelHydrationArgumentValidation private constructor() {
                 hydration: UnderlyingServiceHydration
         ): NadelSchemaValidationError? {
 
-            //need to check null compatibility for inner types (as this is recursive)
-            if (actorFieldArgType.isNonNull && !hydrationSourceFieldType.isNonNull) {
+            //need to check null compatibility
+            val sourceType = remoteArg.remoteArgumentSource.sourceType
+            if (sourceType != ObjectField && actorFieldArgType.isNonNull && !hydrationSourceFieldType.isNonNull) {
```
Comment (sbarker2) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:95]: added `sourceType != ObjectField ` so that it ignores null compatibility for `$source` arguments as that is handled differently by nadel

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+                hydration: UnderlyingServiceHydration,
+                isBatchHydration: Boolean
+        ): NadelSchemaValidationError? {
+            val unwrappedHydrationSourceFieldType = hydrationSourceFieldType.unwrapNonNull()
+            val unwrappedActorFieldArgType = actorFieldArgType.unwrapNonNull()
+
+            //could have ID feed into [ID] (as a possible batch hydration case).
+            // in this case we need to unwrap the list and check types
+            if (isBatchHydration && (!unwrappedHydrationSourceFieldType.isList && unwrappedActorFieldArgType.isList)) {
+                val error = getHydrationInputErrors(
```
Comment (sbarker2) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:31]: Note:

a recurring theme here is:
`val error = getErrors()
if (error != null) { return Error(blah)}`

the reason we cant just simplify this to `return getErrors()` is because we are recursively unwrapping the supplied & actor arg types, but for the error, it is generally more helpful for the message to display the original type, not some inner type that we have unwrapped

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationValidation.kt [hunk 13]
```diff
@@ -1,15 +1,14 @@
 package graphql.nadel.validation
```
Comment (sbarker2) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationValidation.kt:1]: most of this file is just reformatting the code.
the actual code changes are in `getRemoteArgErrors()`

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationValidation.kt [hunk 14]
```diff
@@ -142,50 +139,51 @@ internal class NadelHydrationValidation(
     }
 
     private fun getArgumentErrors(
```
Comment (sbarker2) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationValidation.kt:141]: actual added logic for this file is in this function

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationValidation.kt (hunk 9 above)
Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationValidation.kt:257]: on second thought, this might be fine as its just a single if statement

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+
+        private fun isScalarAssignable(typeToAssign: GraphQLNamedInputType, targetType: GraphQLNamedInputType): Boolean {
+            if (typeToAssign.name == targetType.name) {
+                return true
+            }
+            if (targetType.name == Scalars.GraphQLID.name &&
+                    (typeToAssign.name == Scalars.GraphQLString.name || typeToAssign.name == Scalars.GraphQLInt.name)) {
+                return true
+            }
+            if (targetType.name == Scalars.GraphQLString.name && typeToAssign.name == Scalars.GraphQLID.name) {
```
Comment (sbarker2) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:201]: Allowing ID -> String. see comment here: https://hello.atlassian.net/wiki/spaces/PSRV/pages/2987269760/Analysis+of++Current+Hydration+Arg+Validation+Errors+in+Central+Schema?focusedCommentId=2990608894

**Review by bbakerman — DISMISSED** [review]

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt (hunk 15 above)
Comment (gnawf) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:201]: Not sure I agree with this.

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+                        actorFieldArgType,
+                        actorFieldName
+                )
+            }
+            val unwrappedHydrationSourceFieldType = hydrationSourceFieldType.unwrapNonNull()
+            val unwrappedActorFieldArgType = actorFieldArgType.unwrapNonNull()
+
+            // scalar feed into scalar
+            if (unwrappedHydrationSourceFieldType is GraphQLScalarType && unwrappedActorFieldArgType is GraphQLScalarType) {
+                return validateScalarArg(hydrationSourceFieldType, actorFieldArgType, parent, overallField, remoteArg, actorFieldName)
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:119]: Same comment as before, can you pass in `unwrappedHydrationSourceFieldType` here instead of passing in the raw `GraphQLType` and then unwrapping it inside the function?

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+import graphql.nadel.engine.util.isNonNull
+import graphql.nadel.engine.util.unwrapNonNull
+import graphql.nadel.engine.util.unwrapOne
+import graphql.schema.*
+
+internal class NadelHydrationArgumentValidation private constructor() {
+    companion object {
+        fun validateHydrationInputArg(
+                hydrationSourceFieldType: GraphQLType,
+                actorFieldArgType: GraphQLType,
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:18]:
```suggestion
                actorFieldArgType: GraphQLInputType,
```

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+import graphql.nadel.engine.util.isList
+import graphql.nadel.engine.util.isNonNull
+import graphql.nadel.engine.util.unwrapNonNull
+import graphql.nadel.engine.util.unwrapOne
+import graphql.schema.*
+
+internal class NadelHydrationArgumentValidation private constructor() {
+    companion object {
+        fun validateHydrationInputArg(
+                hydrationSourceFieldType: GraphQLType,
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:17]:
```suggestion
                hydrationSourceFieldType: GraphQLOutputType,
```

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+import graphql.nadel.dsl.RemoteArgumentSource.SourceType.FieldArgument
+import graphql.nadel.dsl.UnderlyingServiceHydration
+import graphql.nadel.engine.util.isList
+import graphql.nadel.engine.util.isNonNull
+import graphql.nadel.engine.util.unwrapNonNull
+import graphql.nadel.engine.util.unwrapOne
+import graphql.schema.*
+
+internal class NadelHydrationArgumentValidation private constructor() {
+    companion object {
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:15]: This shouldn't be a companion object. Refer to the pattern used by the other classes.

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+                        hydrationSourceFieldType,
+                        actorFieldArgType,
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydration,
+                        actorFieldName
+                )
+            }
+            return null
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:88]: This should be baked into the if statements above, it makes it a lot easier to follow.

It should read like

```kotlin
return if () {
  getError()
} else if () {
  getError()
} else {
  getError()
}
```

But right now it's


```kotlin
if () {
  val error = getError()
  if (error != null) {
    return error
  }
} else if () {
  val error = getError()
  if (error != null) {
    return error
  }
} else {
  val error = getError()
  return error
}

return null
```

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+        }
+
+        private fun getHydrationInputErrors(
+                hydrationSourceFieldType: GraphQLType,
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:92]: This isn't strictly correct right? This can be an argument's type right?

I think we should rename this into `sourceType` vs `actorType` (throughout the file).

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+                            remoteArg,
+                            hydrationSourceFieldType,
+                            actorFieldArgType,
+                            actorFieldName
+                    )
+                }
+                return null
+            }
+            // if enums they must be the same
+            else if (unwrappedHydrationSourceFieldType is GraphQLEnumType && unwrappedActorFieldArgType is GraphQLEnumType){
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:144]:
```suggestion
            else if (unwrappedHydrationSourceFieldType is GraphQLEnumType && unwrappedActorFieldArgType is GraphQLEnumType) {
```

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+
+        private fun validateScalarArg(
+                hydrationSourceFieldType: GraphQLType,
+                actorFieldArgType: GraphQLType,
+                parent: NadelServiceSchemaElement,
+                overallField: GraphQLFieldDefinition,
+                remoteArg: RemoteArgumentDefinition,
+                actorFieldName: String
+        ): NadelSchemaValidationError? {
+            if (hydrationSourceFieldType.unwrapNonNull() is GraphQLScalarType && actorFieldArgType.unwrapNonNull() is GraphQLScalarType) {
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:176]: Please add `hydrationSourceFieldType.unwrapNonNull()` as variables so you don't have to manually cast it below.

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 24]
```diff
@@ -0,0 +1,288 @@
...
+                    hydrationSourceFieldInnerType,
+                    actorFieldArgInnerType,
+                    parent,
+                    overallField,
+                    remoteArg,
+                    hydration,
+                    actorFieldName
+            ) != null
+            if (errorExists) {
+                return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:228]: Is there any reason we don't just return the error from the call to `getHydrationInputErrors`?

This seems odd… you're getting an error, checking for its existence, then throwing it away to create a new error?

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt [hunk 25]
```diff
@@ -0,0 +1,288 @@
...
+                    val thisFieldHasError = getHydrationInputErrors(
+                            hydrationType,
+                            actorInnerFieldType,
+                            parent,
+                            overallField,
+                            remoteArg,
+                            hydration,
+                            actorFieldName
+                    ) != null
+                    if (thisFieldHasError) {
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:273]: Same here, why aren't we returning the error from `getHydrationInputErrors` if it exists?

**Review by gnawf — CHANGES_REQUESTED** [review]

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt (hunk 18 above)
Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt:17]: this isnt always a GraphQLOutputType as its sometimes an argument. Will keep as `GraphQLType` and rename `hydrationSourceFieldType` to something more appropriate

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt (hunk 16 above)
Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt]: passing in wrapped type because we want to preserve the wrapped type for the error message to avoid any confusion from the user

Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt]: the only reason we could pass in the unwrapped type previously (in the case that its an object) is because we do not show the type in that one error message

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt (hunk 24 above)
Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt]: the reason for this is because the first error is from checking the inner type,
but I wanted the error to print the type wrapped in the list in the error message, otherwise it may cause confusion for the user

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt (hunk 25 above)
Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt]: similar to the list case, didnt want the error message to say that some random scalar type is not assignable if its wrapped an object (or wrapped multiple layers deep in objects/lists). 

We are effectively transforming the error so the error message can just say the object type that is incompatible, which i think is more useful

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt (hunk 24 above)
Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt]: refer to this comment: https://github.com/atlassian-labs/nadel/pull/463/files/c9298a0f18682717e94b8862371a9c1cabae4cb3#r1357518727

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt (hunk 21 above)
Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt]: Yeah, I am for renaming to `hydrationSourceType`, as it is not always from a field.

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt (hunk 15 above)
Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt]: Sure, we'll have a discussion on the page

File: lib/src/main/java/graphql/nadel/validation/NadelHydrationValidation.kt (hunk 13 above)
Comment (sbarker2) [reply lib/src/main/java/graphql/nadel/validation/NadelHydrationValidation.kt:1]: just reformatted the files I changed using project style, thanks

**Review by gnawf — APPROVED** [review]
